### ISSUE TITLE
Performance with Windows vs Linux

### ISSUE THREAD

**Issue by jchuong** [WPO-Foundation/wptagent#111]
Hi Pat,

## Setup
WPT Server 17.12 release (latest release on Github package, though newer release in AMIs)
wptagent 18.04 image (Linux - ami-236f7f43
Windows (wptagent) - ami-ab5f50cb)
Agents running on us-west-1 on c3.large machines (2 cpu, 3.75GB RAM if I recall correctly).
Browser: Chrome
Connection: Cable (5/1)

I am having some issues with Windows agents taking seemingly 2x~ longer to run tests.

1.  Loading the webpage took 10s vs 20s on linux.
2.  Linux ran 10 runs in 5 minutes, whereas my 10 runs with the windows agent took 16!

I was wondering if there is a known performance difference (maybe 2 cpu isn't enough for Windows, or xvfb is more efficient), and if I have something wrong with my setup (ie, Test run is 10s longer load, but takes a minute to return to WPT and take a new test).

Thanks!

**Comment by pmeenan** [comment]
If you run the tests with debug=1 as a test option they will collect logs from the test running (and processing) which will help identify the issue.  If the loading is as fast (or faster) I wouldn't expect the processing to take that much longer.

That said, Linux is MUCH more stable in general and there are probably still a few rough edges with the agent on Windows (mostly around keeping Windows itself happy and in a good state).

**Comment by jchuong** [comment]
Just to clarify, I want to run a scripted test like

```
debug 1
navigate http://example.com
```
and view the log with `$WPT/viewlog.php?test=$my_test`?

**Comment by pmeenan** [comment]
Sorry, not as a script option, as an API parameters to runtest.php?debug=1....

If you do it through the UI you can go to http://www.myserver.com/?debug=1

In the results to the left of the waterfall (with the timeline links) will be a link to the debug log.

________________________________

**Comment by jchuong** [comment]
Right, thanks.  I got the logs and taking look, nothing 'wrong' seems really stands out too much.

Windows takes a few seconds to start the browser, and a few seconds on the ffmpeg step.

Windows takes 20s from start to finish on the page load, and Linux is like 12s.

Image check takes >8s on Windows, < 1.5s in Linux.

Everything takes a little bit longer on Windows generally. Total time: Windows 49.5s vs Linux 19.2
I did notice a  many more lines of ` Body length: XXX` and `Video file size: 48` on Windows.  Enough to add more seconds to the test.

Also the CPU utilization graph maxes out throughout the test for Windows, and almost never on Linux.  That probably bottle necks the page load itself.

All in all, it's likely Windows needs more CPU power to match with Linux (who would have guessed).

**Comment by pmeenan** [comment]
Generally sounds like the instance size/cpu but going to keep it open just to validate that there aren't optimizations that could be done to improve the optimization checks.

**Comment by jchuong** [comment]
I do believe the (deprecated) wptdriver didn't have as much of a problem if I recall correctly (same instance size, but ran through tests a bit quicker).   I don't have a comparison though and seems like a moot point since wptdriver will be deprecated this year anyways.

**Comment by jchuong** [comment]
Could you create a Windows AMI without RDP? I'd like to try that and see if it keeps it the same as before (I believe some of our users want to test IE occasionally, so Linux only is a hard sell).  Thanks!

**Comment by pmeenan** [comment]
Only available in us-east-1 right now but if it looks good I can copy it to all of the regions and add to the docs: ami-a212a8dd

**Comment by jchuong** [comment]
Thanks Pat, I tried out the AMI but I didn't find a significant difference in their performance against the Windows agent with RDP.  Maybe a second or two off the entire test.  I may just try and stick with Linux agents and just bring up one or two Windows agents for IE when someone needs it.  Thanks again!
